### PR TITLE
[@types/next] Update types for next@7.

### PR DESCRIPTION
**Pull request by joaovieira**
Updated types for Next.js v7.0.x. The majority of the changes are on `next/dynamic` and refactor of the `DocumentProps`.

---

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: Added URLs above type definitions
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/next/head.d.ts [hunk 1]
```diff
@@ -1,8 +1,8 @@
 import * as React from "react";
 
 export function defaultHead(): JSX.Element[];
-export default class extends React.Component {
+export default class Head extends React.Component {
     static canUseDOM: boolean;
-    static peek(): Array<React.ReactElement<any>>;
-    static rewind(): Array<React.ReactElement<any>>;
+    static peek(): React.ReactElement<any>[];
+    static rewind(): React.ReactElement<any>[];
```
Comment (resir014) [inline types/next/head.d.ts]: Will switching these to `React.ReactNodeArray` still work?

**Comment by scottdj92** [comment]
Seems ok to me, but why would the tests fail?

**Review by resir014 — COMMENTED** [review]
Thanks so much for your work! Just one little thing:

File: types/next/head.d.ts (hunk 1 above)
Comment (joaovieira) [reply types/next/head.d.ts]: Just double checked. This is the result of calls to `React.cloneElement` which always returns "sub-classes" of `React.ReactElement` (https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L213). So I think this is correct. I reverted this change and kept the `Array<>` version for non-standard types as per the lint rules.

**Comment by joaovieira** [comment]
One question I have though is: I've seen some types with `package.json` files - do you know if they are used or what do they do? Should we have one and list all the type dependencies (`@types/react`, `@types/react-loadable`, etc.)?

**Comment by scottdj92** [comment]
👌 you're doing god's work son

**Comment by amcasey** [comment]
@resir014 all good?

**Review by resir014 — APPROVED** [review]
Yep, looks good on my end. Thanks for your work @joaovieira!

**Comment by NickBolles** [comment]
I'm running into somme issues getting this to work.

This looks like it doesn't account for Async Document (and all of the other) getInititalProps, should return a promise, and the return type of getInititialProps should probably be the generic type, not the default document type (or it could be a union i guess)

Should I put together a PR for this?

**Comment by scottdj92** [comment]
Submit a PR, and we'll review your work :)

**Comment by joaovieira** [comment]
@NickBolles I see what you're saying and you're right. Those examples are missing in the tests. I've found it hard to make it work though and not sure how to work around it. 

If we change the type to (what seems to be the correct one to me - you can't use the instance templates in the static function):

```tsx
export default class Document<P = DefaultDocumentIProps> extends React.Component<
    P & DocumentProps
> {
    static getInitialProps<IP = DefaultDocumentIProps>(
        context: NextDocumentContext
    ): Promise<IP> | IP;
}
```

The simple example breaks with `DefaultDocumentIProps is no assignable to type 'IP'` 🤷‍♂️ :

screen shot 2018-09-30 at 11 50 46

I'm tempted to relax it `any` to make everyone happy:

```tsx
export default class Document<P = DefaultDocumentIProps> extends React.Component<
    P & DocumentProps
> {
    static getInitialProps(context: NextDocumentContext): any;
}
```

Though, I'm not a TS ninja.. so if someone knows how to do it better feel free to :)

**Pull request merged**